Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPILocalization.mm

```diff
@@ -0,0 +1,324 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "WebExtensionUtilities.h"
+#import <WebKit/_WKWebExtensionLocalization.h>
+
+namespace TestWebKitAPI {
+
+TEST(WKWebExtensionAPILocalization, BuiltInLocalizations)
+{
+    NSDictionary *unlocalizedEnglishDictionary = @{
+        @"current locale": @"__MSG_@@ui_locale__",
+        @"text direction": @"Text direction is __MSG_@@bidi_dir__",
+        @"reversed text direction": @"Text direction is not __MSG_@@bidi_reversed_dir__",
+        @"text start": @"Text starts on the __MSG_@@bidi_start_edge__",
+        @"text end": @"Text direction is __MSG_@@bidi_end_edge__",
+    };
+
+    NSDictionary *expectedLocalizedEnglishDictionary = @{
+        @"current locale": @"en_US",
+        @"text direction": @"Text direction is ltr",
+        @"reversed text direction": @"Text direction is not rtl",
+        @"text start": @"Text starts on the left",
+        @"text end": @"Text direction is right",
+    };
+
+    NSString *uuidString = [NSUUID UUID].UUIDString;
+    _WKWebExtensionLocalization *localization = [[_WKWebExtensionLocalization alloc] initWithRegionalLocalization:@{ } languageLocalization:@{ } defaultLocalization:@{ } withBestLocale:@"en_US" uniqueIdentifier:uuidString];
+    EXPECT_NS_EQUAL([localization localizedDictionaryForDictionary:unlocalizedEnglishDictionary], expectedLocalizedEnglishDictionary);
+
+    EXPECT_WK_STREQ([localization localizedStringForKey:@"@@ui_locale" withPlaceholders:nil], @"en_US");
+    EXPECT_WK_STREQ([localization localizedStringForKey:@"@@bidi_dir" withPlaceholders:nil], @"ltr");
+    EXPECT_WK_STREQ([localization localizedStringForKey:@"@@extension_id" withPlaceholders:nil], uuidString);
+
+    NSDictionary *unlocalizedArabicDictionary = @{
+        @"current locale": @"__MSG_@@ui_locale__",
+        @"text direction": @"Text direction is __MSG_@@bidi_dir__",
+        @"reversed text direction": @"Text direction is not __MSG_@@bidi_reversed_dir__",
+        @"text start": @"Text starts on the __MSG_@@bidi_start_edge__",
+        @"text end": @"Text direction is __MSG_@@bidi_end_edge__",
+    };
+
+    NSDictionary *expectedLocalizedArabicDictionary = @{
+        @"current locale": @"ar_SA",
+        @"text direction": @"Text direction is rtl",
+        @"reversed text direction": @"Text direction is not ltr",
+        @"text start": @"Text starts on the right",
+        @"text end": @"Text direction is left",
+    };
+
+    localization = [[_WKWebExtensionLocalization alloc] initWithRegionalLocalization:@{ } languageLocalization:@{ } defaultLocalization:@{ } withBestLocale:@"ar_SA" uniqueIdentifier:uuidString];
```

Comment:
This is a memory leak, you should either use `RetainPtr` or enable ARC for this file similar to `WebExtensionUtilities.mm`.